### ISSUE TITLE
No way to reload stale songs.

### ISSUE THREAD

**Issue by will-ca** [nukeop/nuclear#1397]
**Platform:**

Flatpak.

**Nuclear version:**

`95cfe0`

**Description of the issue:**

When left for long enough, songs from Youtube stop being playable.

In case of such errors, Nuclear should either automatically invalidate whatever it's cached, and fetch new data, or provide the user with a context menu option to do so.

**Comment by a0kami** [comment]
Hey, what's roughly the waiting time for 'long enough' ?
Do you mean like _the track is left long enough into the queue_ and becomes unplayable ? And if so, is this exclusively affecting the queue ? Does clearing track/re-adding it to queue solves it ?

That'd help a lot to troubleshoot, thanks!

**Comment by nukeop** [comment]
If you remove the track and add it again the URl will be fetched again and should work then.

**Comment by will-ca** [comment]
> If you remove the track and add it again the URl will be fetched again and should work then.

Yeah, what I've been doing is to right click on the track in question, and choose another source, so it fetches a new stream. But mainly this is an annoyance when I have a really long playlist or auto-radio that I want to hear again from the start, but the songs have already expired, in which case doing that one-by-one isn't practical.

**Comment by sferra** [comment]
I have a few suggestions: 
* Don't resolve the URL of too many tracks ahead of time. This strategy will prevent having many stale URLs in a playlist and would also conserve bandwidth. Resolving 1-3 ahead of the current track may be just sufficient. There could even be a setting for this.
* We could store a timestamp along with a resolved URL. This would allow us to determine how long ago a URL was resolved and if it should be considered stale. I suppose that the amount of time would vary per stream provider. If a stale URL can't be used, then we can remove the timestamp and attempt to re-resolve it. This way we can distinguish between "failed because URL was stale" and other types of failures.

**Comment by nukeop** [comment]
It's hard to say when a stream URL goes stale because it can happen any time. A timestamp won't help us. The only way to do it is to figure out that it happened when you try playing, then re-fetch it. Even if you fetch only 1 song ahead, if you put your computer to sleep and come back the next day, the URL will likely be stale.

**Comment by nukeop** [comment]
Nuclear will now re-fetch stream URLs (using the next strea) if it fails when starting playback, as opposed to just removing the tracks from the queue as previously.
